File: core/src/main/scala/com/graphbrain/eco/nodes/ProgNode.scala
package com.graphbrain.eco.nodes

import com.graphbrain.eco.NodeType.NodeType
import com.graphbrain.eco.Contexts

abstract class ProgNode(val lastTokenPos: Int) {
  def ntype: NodeType

  def stringValue(ctxts: Contexts): Unit = {} // error
  def numberValue(ctxts: Contexts): Unit = {} // error
  def booleanValue(ctxts: Contexts): Unit = {} // error
  def wordsValue(ctxts: Contexts): Unit = {} // error
  def vertexValue(ctxts: Contexts): Unit = {} // error
  def verticesValue(ctxts: Contexts): Unit = {} // error

  protected def error(msg: String) = println(msg)

  protected def typeError() = error("type error")
}

File: core/src/main/java/com/graphbrain/eco/nodes/ProgNode.java
package com.graphbrain.eco.nodes;

import com.graphbrain.eco.Contexts;
import com.graphbrain.eco.NodeType;

public abstract class ProgNode {

    private int lastTokenPos;

    public ProgNode(int lastTokenPos) {
        this.lastTokenPos = lastTokenPos;
    }

    public abstract NodeType ntype();

    public void stringValue(Contexts ctxts) {} // error
    public void numberValue(Contexts ctxts) {} // error
    public void booleanValue(Contexts ctxts) {} // error
    public void wordsValue(Contexts ctxts) {} // error
    public void vertexValue(Contexts ctxts) {} // error
    public void verticesValue(Contexts ctxts) {} // error

    public void value(Contexts ctxts) {
        switch(ntype()) {
            case Boolean:
                booleanValue(ctxts);
                break;
            case Number:
                numberValue(ctxts);
                break;
            case Words:
                wordsValue(ctxts);
                break;
            case String:
                stringValue(ctxts);
                break;
            case Vertex:
                vertexValue(ctxts);
                break;
        }
    }

    protected void error(String msg) {
        System.out.println(msg);
    }

    protected void typeError() {
        error("type error");
    }

    public int getLastTokenPos() {
        return lastTokenPos;
    }
}